lib/Onyx.d.ts

```diff
@@ -0,0 +1,337 @@
+ *
+ * It's very similar to `KeyValueMapping` but this type accepts using `null` as well.
+ */
+type NullableKeyValueMapping = {
+    [TKey in OnyxKey]: OnyxRecord<KeyValueMapping[TKey]>;
+};
+
+/**
+ * Represents the base options used in `Onyx.connect()` method.
+ */
+type BaseConnectOptions<TKey extends OnyxKey> = {
+    statePropertyName?: string;
+    withOnyxInstance?: Component;
+    initWithStoredValues?: boolean;
+    selector?: Selector<TKey, OnyxRecord<KeyValueMapping[TKey]>>;
+};
+
+/**
+ * Represents the options used in `Onyx.connect()` method.
+ * The type is built from `BaseConnectOptions` and extended to handle key/callback related options.
+ * It includes two different forms, depending on whether we are waiting for a collection callback or not.
+ *
+ * If `waitForCollectionCallback` is `true`, it expects `key` to be a Onyx collection key and `callback` will pass `value` as an `OnyxCollectionRecords`.
+ *
+ * If `waitForCollectionCallback` is `false` or not specified, the `key` can be any Onyx key and `callback` will pass `value` as an `OnyxRecord`.
+ */
+type ConnectOptions<TKey extends OnyxKey> = BaseConnectOptions<TKey> &
+    (
+        | {
+              key: TKey extends CollectionKey ? TKey : never;
+              callback?: (value: OnyxCollectionRecords<KeyValueMapping[TKey]>, key?: TKey) => void;
+              waitForCollectionCallback: true;
+          }
+        | {
+              key: TKey;
+              callback?: (value: OnyxRecord<KeyValueMapping[TKey]>, key?: TKey) => void;
+              waitForCollectionCallback?: false;
+          }
+    );
+
+/**
+ * Represents a mapping between Onyx collection keys and their respective values.
+ * It helps to enforce that a Onyx collection key should not be without suffix (e.g. should always be of the form `${TKey}${string}`),
+ * and to map each Onyx collection key with suffix to a value of type `TValue`.
+ */
+type Collection<TKey extends Key, TMap, TValue> = {
+    [MapK in keyof TMap]: MapK extends `${TKey}${string}`
+        ? MapK extends `${TKey}`
+            ? never // forbids empty id
+            : TValue
+        : never;
+};
+
+/**
+ * Represents different kinds of updates that can be passed to `Onyx.update()` method. It is a discriminated union of
+ * different update methods (`SET`, `MERGE`, `MERGE_COLLECTION`), each with their own key and value structure.
+ */
+type OnyxUpdate<TKey extends OnyxKey> =
+    | {
+          onyxMethod: typeof METHOD.SET;
+          key: TKey;
+          value: OnyxRecord<KeyValueMapping[TKey]>;
+      }
+    | {
+          onyxMethod: typeof METHOD.MERGE;
+          key: TKey;
+          value: PartialDeep<KeyValueMapping[TKey]>;
+      }
+    | {
+          [TKey in CollectionKey]: {
+              onyxMethod: typeof METHOD.MERGE_COLLECTION;
+              key: TKey;
+              value: Record<`${TKey}${string}`, PartialDeep<KeyValueMapping[TKey]>>;
```

Comment:
Maybe with `CollectionKeyBase` we can ensure that keys in Record are actually `CollectionKey` and not `CollectionKeyBase`, wdyt? @fabioh8010